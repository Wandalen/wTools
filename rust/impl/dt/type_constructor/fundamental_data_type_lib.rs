#![ doc( html_logo_url = "https://raw.githubusercontent.com/Wandalen/wTools/master/asset/img/logo_v3_trans_square.png" ) ]
#![ warn( rust_2018_idioms ) ]
#![ warn( missing_debug_implementations ) ]
#![ warn( missing_docs ) ]

//!
//! Fundamental data types and type constructors, like Single, Pair, Many.
//!

#![ doc = include_str!( concat!( env!( "CARGO_MANIFEST_DIR" ), "/Readme.md" ) ) ]

#[ doc( inline ) ]
pub use ::type_constructor::*;
