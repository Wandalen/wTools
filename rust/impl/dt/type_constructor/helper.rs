/// Internal namespace.
pub( crate ) mod private
{
  use crate::exposed::*;

  ///
  /// Generate cod only if feature::make is enabled.
  ///
  /// Do not use manually.
  ///

  #[ cfg( feature = "make" ) ]
  #[ macro_export ]
  macro_rules! _if_make
  {
    ( $( $Rest : tt )* ) =>
    {
      $( $Rest )*
    };
  }

  #[ cfg( not( feature = "make" ) ) ]
  #[ macro_export ]
  macro_rules! _if_make
  {
  }

  pub use _if_make;
}

/// Protected namespace of the module.
pub mod protected
{
  pub use super::orphan::*;
}

pub use protected::*;

/// Orphan namespace of the module.
pub mod orphan
{
  pub use super::exposed::*;
}

/// Exposed namespace of the module.
pub mod exposed
{
  pub use super::prelude::*;
}

pub use exposed::*;

/// Prelude to use essentials: `use my_module::prelude::*`.
pub mod prelude
{
  pub use super::private::
  {
    _if_make,
  };
}