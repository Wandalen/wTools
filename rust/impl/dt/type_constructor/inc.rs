
/// Type constructor of many.
// #[ cfg( feature = "types" ) ]
pub mod many;
/// Type constructor of pair.
// #[ cfg( feature = "types" ) ]
pub mod pair;
/// Type constructor of single.
// #[ cfg( feature = "types" ) ]
pub mod single;
/// Type constructors.
// #[ cfg( feature = "types" ) ]
pub mod types;

/// Macro helpers.
pub mod helper;
/// Generic traits.
pub mod traits;
/// Variadic constructor.
#[ cfg( feature = "make" ) ]
pub mod make;

/// Protected namespace of the module.
pub mod protected
{
  #[ doc( inline ) ]
  pub use super::orphan::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::many::orphan::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::pair::orphan::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::single::orphan::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::types::orphan::*;
  #[ doc( inline ) ]
  pub use super::helper::orphan::*;
  #[ doc( inline ) ]
  pub use super::traits::orphan::*;
  #[ doc( inline ) ]
  #[ cfg( feature = "make" ) ]
  pub use super::make::orphan::*;
}

#[ doc( inline ) ]
pub use protected::*;

/// Orphan namespace of the module.
pub mod orphan
{
  pub use super::exposed::*;
}

/// Exposed namespace of the module.
pub mod exposed
{
  #[ doc( inline ) ]
  pub use super::prelude::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::many::exposed::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::pair::exposed::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::single::exposed::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::types::exposed::*;
  #[ doc( inline ) ]
  pub use super::helper::exposed::*;
  #[ doc( inline ) ]
  pub use super::traits::exposed::*;
  #[ doc( inline ) ]
  #[ cfg( feature = "make" ) ]
  pub use super::make::exposed::*;
}

/// Prelude to use essentials: `use my_module::prelude::*`.
pub mod prelude
{
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::many::prelude::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::pair::prelude::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::single::prelude::*;
  #[ doc( inline ) ]
  // #[ cfg( feature = "types" ) ]
  pub use super::types::prelude::*;
  #[ doc( inline ) ]
  pub use super::helper::prelude::*;
  #[ doc( inline ) ]
  pub use super::traits::prelude::*;
  #[ doc( inline ) ]
  #[ cfg( feature = "make" ) ]
  pub use super::make::prelude::*;
}
