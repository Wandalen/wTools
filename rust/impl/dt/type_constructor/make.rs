/// Internal namespace.
#[ cfg( feature = "make" ) ]
pub( crate ) mod private
{

  ///
  /// Constructor without arguments.
  ///

  pub trait Make0
  where
    Self : Sized,
  {
    /// Constructor without arguments.
    fn make() -> Self
    {
      Self::make_0()
    }
    /// Constructor without arguments.
    fn make_0() -> Self;
  }

  ///
  /// Constructor with single argument.
  ///

  pub trait Make1< Arg >
  where
    Self : Sized,
  {
    /// Constructor without arguments.
    fn make( arg : Arg ) -> Self
    {
      Self::make_1( arg )
    }
    /// Constructor without arguments.
    fn make_1( arg : Arg ) -> Self;
  }

  ///
  /// Constructor with two arguments.
  ///

  pub trait Make2< Arg1, Arg2 >
  where
    Self : Sized,
  {
    /// Constructor with two arguments.
    fn make( arg1 : Arg1, arg2 : Arg2 ) -> Self
    {
      Self::make_2( arg1, arg2 )
    }
    /// Constructor with two arguments.
    fn make_2( arg1 : Arg1, arg2 : Arg2 ) -> Self;
  }

  ///
  /// Constructor with three arguments.
  ///

  pub trait Make3< Arg1, Arg2, Arg3 >
  where
    Self : Sized,
  {
    /// Constructor with three arguments.
    fn make( arg1 : Arg1, arg2 : Arg2, arg3 : Arg3 ) -> Self
    {
      Self::make_3( arg1, arg2, arg3 )
    }
    /// Constructor with three arguments.
    fn make_3( arg1 : Arg1, arg2 : Arg2, arg3 : Arg3 ) -> Self;
  }

  ///
  /// Variadic constructor.
  ///
  /// Implement traits [Make0], [Make1] up to MakeN to provide the interface to construct your structure with a different set of arguments.
  /// In this example structure, Struct1 could be constructed either without arguments, with a single argument, or with two arguments.
  /// - Constructor without arguments fills fields with zero.
  /// - Constructor with a single argument sets both fields to the value of the argument.
  /// - Constructor with 2 arguments set individual values of each field.
  ///
  /// ```rust
  /// #[ cfg( feature = "make" ) ]
  /// {
  ///   use type_constructor::prelude::*;
  ///
  ///   #[ derive( Debug, PartialEq ) ]
  ///   struct Struct1
  ///   {
  ///     a : i32,
  ///     b : i32,
  ///   }
  ///
  ///   impl Make0 for Struct1
  ///   {
  ///     fn make_0() -> Self
  ///     {
  ///       Self { a : 0, b : 0 }
  ///     }
  ///   }
  ///
  ///   impl Make1< i32 > for Struct1
  ///   {
  ///     fn make_1( val : i32 ) -> Self
  ///     {
  ///       Self { a : val, b : val }
  ///     }
  ///   }
  ///
  ///   impl Make2< i32, i32 > for Struct1
  ///   {
  ///     fn make_2( val1 : i32, val2 : i32 ) -> Self
  ///     {
  ///       Self { a : val1, b : val2 }
  ///     }
  ///   }
  ///
  ///   let got : Struct1 = make!();
  ///   let exp = Struct1{ a : 0, b : 0 };
  ///   assert_eq!( got, exp );
  ///
  ///   let got : Struct1 = make!( 13 );
  ///   let exp = Struct1{ a : 13, b : 13 };
  ///   assert_eq!( got, exp );
  ///
  ///   let got : Struct1 = make!( 1, 3 );
  ///   let exp = Struct1{ a : 1, b : 3 };
  ///   assert_eq!( got, exp );
  /// }
  ///
  /// ```
  ///
  /// ### To add to your project
  ///
  /// ``` shell
  /// cargo add type_constructor
  /// ```
  ///
  /// ## Try out from the repository
  ///
  /// ``` shell test
  /// git clone https://github.com/Wandalen/wTools
  /// cd wTools
  /// cd sample/rust/type_constructor_trivial_sample
  /// cargo run
  /// ```

  #[ macro_export ]
  macro_rules! make
  {

    (
      $(,)?
    )
    =>
    {
      $crate::Make0::make_0();
    };

    (
      $Arg1 : expr $(,)?
    )
    =>
    {
      $crate::Make1::make_1( $Arg1 );
    };

    (
      $Arg1 : expr, $Arg2 : expr $(,)?
    )
    =>
    {
      $crate::Make2::make_2( $Arg1, $Arg2 );
    };

    (
      $Arg1 : expr, $Arg2 : expr, $Arg3 : expr $(,)?
    )
    =>
    {
      $crate::Make3::make_3( $Arg1, $Arg2, $Arg3 );
    };

    (
      $( $Rest : tt )+
    )
    =>
    {
      compile_error!
      (
        concat!
        (
          "Variadic constructor supports up to 3 arguments.\n",
          "Open an issue if you need more.\n",
          "You passed:\n",
          stringify!
          (
            make!( $( $Rest )+ )
          )
        )
      );
    };

  }

  pub use make;
}

/// Protected namespace of the module.
pub mod protected
{
  #[ doc( inline ) ]
  pub use super::orphan::*;
}

#[ doc( inline ) ]
pub use protected::*;

/// Orphan namespace of the module.
pub mod orphan
{
  #[ doc( inline ) ]
  pub use super::exposed::*;
}

/// Exposed namespace of the module.
pub mod exposed
{
  #[ doc( inline ) ]
  pub use super::prelude::*;
}

#[ doc( inline ) ]
pub use exposed::*;

/// Prelude to use essentials: `use my_module::prelude::*`.
pub mod prelude
{
  #[ cfg( feature = "make" ) ]
  #[ doc( inline ) ]
  pub use super::private::
  {

    Make0,
    Make1,
    Make2,
    Make3,

    make,

  };
}
