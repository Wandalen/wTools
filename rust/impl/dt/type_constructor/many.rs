/// Internal namespace.
pub( crate ) mod private
{
  use crate::exposed::*;

  ///
  /// Type constructor of many.
  ///
  /// Should not be used directly. Instead use macro [crate::types!].
  ///

  #[ macro_export ]
  macro_rules! _many
  {

    // many Many : < T >;

    (
      $( #[ $Meta : meta ] )*
      many $Name : ident :
      < $ParamName : ident $( : $ParamTy1x1 : ident $( :: $ParamTy1xN : ident )* $( + $ParamTy2 : path )* )? >
      $( ; $( $Rest : tt )* )?
    )
    =>
    {
      $( #[ $Meta ] )*
      pub struct $Name
      < $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? >
      ( pub std::vec::Vec< $ParamName > );

      impl< $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? > core::ops::Deref
      for $Name
      < $ParamName >
      {
        type Target = std::vec::Vec< $ParamName >;
        fn deref( &self ) -> &Self::Target
        {
          &self.0
        }
      }

      impl< $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? > core::ops::DerefMut
      for $Name
      < $ParamName >
      {
        fn deref_mut( &mut self ) -> &mut Self::Target
        {
          &mut self.0
        }
      }

      impl< $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? >
      From< $ParamName >
      for $Name
      < $ParamName >
      {
        fn from( src : $ParamName ) -> Self
        {
          Self( vec![ src ] )
        }
      }

      impl< $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? >
      From< ( $ParamName, ) >
      for $Name
      < $ParamName >
      {
        fn from( src : ( $ParamName, ) ) -> Self
        {
          Self( vec![ src.0 ] )
        }
      }

      impl< $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )?, const N : usize >
      From< [ $ParamName ; N ] >
      for $Name
      < $ParamName >
      where
        $ParamName : Clone,
      {
        fn from( src : [ $ParamName ; N ] ) -> Self
        {
          Self( std::vec::Vec::from( src ) )
        }
      }

      impl< $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? >
      From< &[ $ParamName ] >
      for $Name
      < $ParamName >
      where
        $ParamName : Clone,
      {
        fn from( src : &[ $ParamName ] ) -> Self
        {
          Self( std::vec::Vec::from( src ) )
        }
      }

      impl< $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? >
      AsSlice< $ParamName >
      for $Name < $ParamName >
      {
        fn as_slice( &self ) -> &[ $ParamName ]
        {
          &self[ .. ]
        }
      }

      $crate::_if_make!
      {

        impl< $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? >
        Make0
        for $Name < $ParamName >
        {
          fn make_0() -> Self
          {
            Self( std::vec::Vec::new() )
          }
        }

        impl< $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? >
        Make1< $ParamName >
        for $Name < $ParamName >
        {
          fn make_1( _0 : $ParamName ) -> Self
          {
            Self( vec![ _0 ] )
          }
        }

        impl< $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? >
        Make2< $ParamName, $ParamName >
        for $Name < $ParamName >
        {
          fn make_2( _0 : $ParamName, _1 : $ParamName ) -> Self
          {
            Self( vec![ _0, _1 ] )
          }
        }

        impl< $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? >
        Make3< $ParamName, $ParamName, $ParamName >
        for $Name < $ParamName >
        {
          fn make_3( _0 : $ParamName, _1 : $ParamName, _2 : $ParamName ) -> Self
          {
            Self( vec![ _0, _1, _2 ] )
          }
        }

      }

      $crate::types!{ $( $( $Rest )* )? }
    };

    // many Many : < T1, ... >;

    (
      $( #[ $Meta : meta ] )*
      many $Name : ident :
      < $ParamName : ident $( : $ParamTy1x1 : ident $( :: $ParamTy1xN : ident )* $( + $ParamTy2 : path )* )? ,
      $( $Rest : tt )*
    )
    =>
    {
      compile_error!
      (
        concat!
        (
          "Parametrized element should be many, because Many has only one element\n",
          stringify!
          (
            $( #[ $Meta ] )*
            $Name :
            < $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ,
            $( $Rest )*
          )
        )
      );
    };

    // many Many : Element< T1, T2, ... >;

    (
      $( #[ $Meta : meta ] )*
      many $Name : ident : $TypeSplit1 : ident $( :: $TypeSplitN : ident )*
      $( < $( $ParamName : ident $( : $ParamTy1x1 : ident $( :: $ParamTy1xN : ident )* $( + $ParamTy2 : path )* )? ),* > )?
      $( ; $( $Rest : tt )* )?
    )
    =>
    {
      $( #[ $Meta ] )*
      pub struct $Name
      $( < $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* > )?
      ( pub std::vec::Vec< $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? > );

      impl
      $( < $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* > )?
      core::ops::Deref
      for $Name
      $( < $( $ParamName ),* > )?
      {
        type Target = std::vec::Vec< $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? >;
        fn deref( &self ) -> &Self::Target
        {
          &self.0
        }
      }

      impl
      $( < $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* > )?
      core::ops::DerefMut
      for $Name
      $( < $( $ParamName ),* > )?
      {
        fn deref_mut( &mut self ) -> &mut Self::Target
        {
          &mut self.0
        }
      }

      impl
      $( < $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* > )?
      From
      < $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? >
      for $Name
      $( < $( $ParamName ),* > )?
      {
        fn from( src : $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? ) -> Self
        {
          Self( vec![ src ] )
        }
      }

      impl
      $( < $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* > )?
      From
      < ( $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? , ) >
      for $Name
      $( < $( $ParamName ),* > )?
      {
        fn from( src : ( $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? , ) ) -> Self
        {
          Self( vec![ src.0 ] )
        }
      }

      impl
      < $( $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? , )* )? const N : usize >
      From
      < [ $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? ; N ] >
      for $Name
      $( < $( $ParamName ),* > )?
      where
        $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? : Clone,
      {
        fn from( src : [ $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? ; N ] ) -> Self
        {
          Self( std::vec::Vec::from( src ) )
        }
      }

      impl
      $( < $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* > )?
      From
      < &[ $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? ] >
      for $Name
      $( < $( $ParamName ),* > )?
      where
        $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? : Clone,
      {
        fn from( src : &[ $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? ] ) -> Self
        {
          debug_assert_eq!( src.len(), 1 ); // xxx : add test with long slice
          Self( std::vec::Vec::from( src ) )
        }
      }

      impl
      $( < $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* > )?
      AsSlice< $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? >
      for
      $Name $( < $( $ParamName ),* > )?
      where
        $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? : Clone,
      {
        fn as_slice( &self ) -> &[ $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? ]
        {
          &self[ .. ]
        }
      }

      $crate::_if_make!
      {

        // #[ cfg( feature = "make" ) ]
        impl
        $( < $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* > )?
        Make0
        for
        $Name $( < $( $ParamName ),* > )?
        {
          fn make_0() -> Self
          {
            Self( std::vec::Vec::< $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? >::new() )
          }
        }

        // #[ cfg( feature = "make" ) ]
        impl
        $( < $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* > )?
        Make1< $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )? >
        for
        $Name $( < $( $ParamName ),* > )?
        {
          fn make_1
          (
            _0 : $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )?,
          )
          -> Self
          {
            Self( vec![ _0 ] )
          }
        }

        // #[ cfg( feature = "make" ) ]
        impl
        $( < $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* > )?
        Make2
        <
          $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )?,
          $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )?,
        >
        for
        $Name $( < $( $ParamName ),* > )?
        {
          fn make_2
          (
            _0 : $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )?,
            _1 : $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )?,
          )
          -> Self
          {
            Self( vec![ _0, _1 ] )
          }
        }

        // #[ cfg( feature = "make" ) ]
        impl
        $( < $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* > )?
        Make3
        <
          $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )?,
          $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )?,
          $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )?,
        >
        for
        $Name $( < $( $ParamName ),* > )?
        {
          fn make_3
          (
            _0 : $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )?,
            _1 : $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )?,
            _2 : $TypeSplit1 $( :: $TypeSplitN )* $( < $( $ParamName ),* > )?,
          )
          -> Self
          {
            Self( vec![ _0, _1, _2 ] )
          }
        }

      }

      $crate::types!{ $( $( $Rest )* )? }
    };

  }

  types!
  {

    ///
    /// Type constructor to wrap a another type into a tuple.
    ///

    // xxx : sample

    #[ derive( Debug, Clone, PartialEq, Eq, Default ) ]
    many Many : < T >;

  }

  pub use _many;
}

/// Protected namespace of the module.
pub mod protected
{
  pub use super::orphan::*;
}

pub use protected::*;

/// Orphan namespace of the module.
pub mod orphan
{
  pub use super::exposed::*;
}

/// Exposed namespace of the module.
pub mod exposed
{
  pub use super::prelude::*;
}

pub use exposed::*;

/// Prelude to use essentials: `use my_module::prelude::*`.
pub mod prelude
{
  pub use super::private::
  {

    _many,
    Many,

  };
}
