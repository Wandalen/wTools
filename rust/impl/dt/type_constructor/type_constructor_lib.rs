
#![ cfg_attr( not( feature = "use_std" ), no_std ) ]
#![ doc( html_logo_url = "https://raw.githubusercontent.com/Wandalen/wTools/master/asset/img/logo_v3_trans_square.png" ) ]
#![ doc( html_favicon_url = "https://raw.githubusercontent.com/Wandalen/wTools/alpha/asset/img/logo_v3_trans_square_icon_small_v2.ico")]
#![ doc( html_root_url = "https://docs.rs/type_constructor/latest/type_constructor/")]
#![ warn( rust_2018_idioms ) ]
#![ warn( missing_debug_implementations ) ]
#![ warn( missing_docs ) ]

//!
//! Fundamental data types and type constructors, like Single, Pair, Many.
//!

#![ doc = include_str!( concat!( env!( "CARGO_MANIFEST_DIR" ), "/", "Readme.md" ) ) ]

// #![ no_std ]

// #[ cfg( not( feature = "use_std" ) ) ]
// extern crate core as std;
// #[ cfg( all( not( feature = "use_std" ), feature = "use_alloc" ) ) ]
// extern crate alloc;

#[ path = "./inc.rs" ]
mod inc;
#[ doc( inline ) ]
pub use inc::*;
