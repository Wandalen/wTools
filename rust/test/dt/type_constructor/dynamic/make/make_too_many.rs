use type_constructor as TheModule;
use TheModule::prelude::*;

fn main()
{
  let x = make!( 0, 1, 2, 3 );
}

/* xxx : does not work properly currently */