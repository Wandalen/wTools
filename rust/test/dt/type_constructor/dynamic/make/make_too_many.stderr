error: Variadic constructor supports up to 3 arguments.
       Open an issue if you need more.
       You passed:
       make! (0, 1, 2, 3)
 --> ../../../rust/test/dt/type_constructor/dynamic/make/make_too_many.rs:6:11
  |
6 |   let x = make!( 0, 1, 2, 3 );
  |           ^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `make` (in Nightly builds, run with -Z macro-backtrace for more info)
